module/documents/activity/attack.mjs

```diff
@@ -21,7 +22,192 @@ export default class AttackActivity extends ActivityMixin(AttackActivityData) {
       type: "attack",
       img: "systems/dnd5e/icons/svg/activity/attack.svg",
       title: "DND5E.ATTACK.Title.one",
-      sheetClass: AttackSheet
+      sheetClass: AttackSheet,
+      usage: {
+        actions: {
+          rollAttack: AttackActivity.#rollAttack
+        }
+      }
     }, { inplace: false })
   );
+
+  /* -------------------------------------------- */
+  /*  Activation                                  */
+  /* -------------------------------------------- */
+
+  /** @override */
+  _usageChatButtons() {
+    return [{
+      label: game.i18n.localize("DND5E.Attack"),
+      icon: '<i class="dnd5e-icon" data-src="systems/dnd5e/icons/svg/trait-weapon-proficiencies.svg" inert></i>',
+      dataset: {
+        action: "rollAttack"
+      }
+    }];
+  }
+
+  /* -------------------------------------------- */
+  /*  Rolling                                     */
+  /* -------------------------------------------- */
+
+  /**
+   * Perform an attack roll.
+   * @param {D20RollProcessConfiguration} config     Configuration information for the roll.
+   * @param {BasicRollDialogConfiguration} dialog    Configuration for the roll dialog.
+   * @param {BasicRollMessageConfiguration} message  Configuration for the roll message.
+   * @returns {Promise<D20Roll[]|void>}
+   */
+  async rollAttack(config={}, dialog={}, message={}) {
+    const { parts, data } = this.getAttackData();
+    const targets = this.constructor.getTargetDescriptors();
+
+    let ammoUpdate = [];
+    // TODO: Handle ammunition consumption
+    // const consume = this.system.consume;
+    // const ammo = this.hasAmmo ? this.actor.items.get(consume.target) : null;
+    // if ( ammo ) {
+    //   const q = ammo.system.quantity;
+    //   const consumeAmount = consume.amount ?? 0;
+    //   if ( q && (q - consumeAmount >= 0) ) {
+    //     title += ` [${ammo.name}]`;
+    //   }
+    //
+    //   // Get pending ammunition update
+    //   const usage = this._getUsageUpdates({consumeResource: true});
+    //   if ( usage === false ) return null;
+    //   ammoUpdate = usage.resourceUpdates ?? [];
+    // }
+
+    const rollConfig = foundry.utils.mergeObject({
+      elvenAccuracy: this.actor?.getFlag("dnd5e", "elvenAccuracy")
+        && CONFIG.DND5E.characterFlags.elvenAccuracy.abilities.includes(this.ability),
+      halflingLucky: this.actor?.getFlag("dnd5e", "halflingLucky")
+    }, config);
+    rollConfig.origin = this;
+    rollConfig.rolls = [{
+      parts, data,
+      options: {
+        criticalSuccess: this.criticalThreshold,
+        target: targets.length === 1 ? targets[0].ac : undefined
+      }
+    }].concat(config.rolls ?? []);
+
+    const dialogConfig = foundry.utils.mergeObject({
+      configure: true,
+      options: {
+        width: 400,
+        top: config.event ? config.event.clientY - 80 : null,
+        left: window.innerWidth - 710
+      }
+    }, dialog);
```

Comment:
Would be nice to have this factored out somewhere as well, assuming other activities are going to be doing the same thing. If it's going to be part of a refactored `D20Roll` though, then no worries.

Reply:
Yeah, we can add something to the `BasicRollConfigurationDialog` static constructor to handle this.